hbase-server/src/main/java/org/apache/hadoop/hbase/master/webapp/MetaBrowser.java

```diff
@@ -0,0 +1,360 @@
+import org.apache.hadoop.hbase.filter.PrefixFilter;
+import org.apache.hadoop.hbase.filter.SingleColumnValueFilter;
+import org.apache.hadoop.hbase.master.RegionState;
+import org.apache.hadoop.hbase.util.Bytes;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.apache.hbase.thirdparty.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.hbase.thirdparty.io.netty.handler.codec.http.QueryStringEncoder;
+
+/**
+ * A support class for the "Meta Entries" section in
+ * {@code resources/hbase-webapps/master/table.jsp}.
+ */
+@InterfaceAudience.Private
+public class MetaBrowser implements Iterable<RegionReplicaInfo> {
+  private static final Logger logger = LoggerFactory.getLogger(MetaBrowser.class);
+
+  public static final String NAME_PARAM = "name";
+  public static final String SCAN_LIMIT_PARAM = "scan_limit";
+  public static final String SCAN_REGION_STATE_PARAM = "scan_region_state";
+  public static final String SCAN_START_PARAM = "scan_start";
+  public static final String SCAN_TABLE_PARAM = "scan_table";
+
+  public static final int SCAN_LIMIT_DEFAULT = 10;
+  public static final int SCAN_LIMIT_MAX = 10_000;
+
+  private final AsyncConnection connection;
+  private final HttpServletRequest request;
+  private final ExecutorService pool;
+  private final List<String> errorMessages;
+  private final String name;
+  private final Integer scanLimit;
+  private final RegionState.State scanRegionState;
+  private final byte[] scanStart;
+  private final TableName scanTable;
+
+  public MetaBrowser(final AsyncConnection connection, final HttpServletRequest request) {
+    this.connection = connection;
+    this.request = request;
+    this.pool = buildThreadPool();
+    this.errorMessages = new LinkedList<>();
+    this.name = resolveName(request);
+    this.scanLimit = resolveScanLimit(request);
+    this.scanRegionState = resolveScanRegionState(request);
+    this.scanStart = resolveScanStart(request);
+    this.scanTable = resolveScanTable(request);
+  }
+
+  public List<String> getErrorMessages() {
+    return errorMessages;
+  }
+
+  public String getName() {
+    return name;
+  }
+
+  public Integer getScanLimit() {
+    return scanLimit;
+  }
+
+  public byte[] getScanStart() {
+    return scanStart;
+  }
+
+  public RegionState.State getScanRegionState() {
+    return scanRegionState;
+  }
+
+  public TableName getScanTable() {
+    return scanTable;
+  }
+
+  @Override
+  public Iterator<RegionReplicaInfo> iterator() {
+    return limitIterator();
+  }
+
+  public LimitIterator<RegionReplicaInfo> limitIterator() {
+    logger.debug("initiating meta scan, {}", this);
+
+    final AsyncTable<ScanResultConsumer> asyncTable =
+      connection.getTable(TableName.META_TABLE_NAME, pool);
+    // TODO: buffering the entire result set seems unnecessary.
+    final List<RegionReplicaInfo> results = asyncTable.scanAll(buildScan()).join()
```

Comment:
Scan all each time we page?

Reply:
MetaTable Visitor no good to you in here?

Reply:
> Scan all each time we page?

Scan all of meta? No. Scanning starts at the row represented by `SCAN_START_PARAM` and is limited by `SCAN_LIMIT_PARAM` and the other filter parameters. See the `TestMetaBrowser#paginate*` tests.

> MetaTable Visitor no good to you in here?

Is that a class? I figured out how to do away with `scanAll`; moving to the old fashioned `ResultScanner`.